proposals/inline-classes.md

```diff
@@ -0,0 +1,296 @@
+
+### Type mapping on JVM
+
+Depending on the underlying type of inline class and its declaration site, inline class type can be mapped to the underlying type or to the 
+boxed type.  
+
+| Underlying Type \ Declaration Site | Not-null | Nullable | Generic |
+| --------------------------------------------- | --------------- | ---------------- | ---------------- |
+| **Reference type** | Underlying type | Underlying type | Boxed type |
+| **Nullable reference type** | Underlying type  | Boxed type | Boxed type |
+| **Primitive type** | Underlying type  | Boxed type | Boxed type |
+| **Nullable primitive type** | Underlying type  | Boxed type | Boxed type |
+
+For example, from this table it follows that nullable inline class which is based on reference type is mapped to the underlying type:
+```kotlin
+inline class Name(val s: String)
+
+fun foo(n: Name?) { ... } // `Name?` here is mapped to String
+```
+
+Also, if inline class type is used in generic position, then its boxed type will be used:
+```
+// Kotlin: sample.kt
+
+inline class Name(val s: String)
+
+fun generic(names: List<Name>) {} // generic signature will have `List<Name>` as for parameters type
+fun simple(name: Name) {}
+
+// Java
+class Test {
+    void test() {
+        String name = SampleKt.simple();
+        List<Name> ls = Samplekt.generic(); // from Java POV it's List<Name>, not List<String>
+    }
+}
+```
+
+
+#### Generic inline class mapping
+
+Consider the following sample:
+```kotlin
+inline class Generic<T>(val x: T)
+
+fun foo(g: Generic<String>) {}
+```
+
+Now, type `Generic<String>` can be mapped either to `java.lang.String` or to `java.lang.Object`. 
+To make the whole rule more consistent, currently we propose to map `Generic<SomeType>` always to `java.lang.Object`, 
+i.e. we'll map upper bound of type parameter.
+
+* Sidenote: maybe it's worth to consider inline classes with reified generics:
+    ```kotlin
+    inline class Reified<reified T>(val x: T)
+    
+    fun foo(a: Reified<Int>, b: Reified<String>) // a has type `Int`, b has type `String`
+    ``` 
+
+Generic inline classes with underlying value not of type that defined by type parameter are mapped as usual generics:
+```kotlin
+inline class AsList<T>(val ls: List<T>)
+
+fun foo(param: AsList<String>) {}
+```
+
+In JVM signature `param` will have type `java.util.List`, 
+but in generic signature it will be `java.util.List<java.lang.String>` 
+
+## Methods from `kotlin.Any`
+
+Inline classes are indirectly inherited from `Any`, i.e. they can be assigned to a value of type `Any`, but only through boxing.
+
+Methods from `Any` (`toString`, `hashCode`, `equals`) can be useful for a user-defined inline classes and therefore should be customizable. 
+Methods `toString` and `hashCode` can be overridden as usual methods from `Any`. For method `equals` we're going to introduce new operator 
```

Comment:
Wait, how will overridding `toString()`, `hashCode()` and `equals()` work? 

```kotlin
inline class UserId(private val id: String) {
    override fun toString() = "UserId($id)"
}

fun test(userId: UserId) {
    System.out.println(userId) // ?
}
```

It'll require a wrapper class which is what `inline class` is trying to remove

Reply:
@artem-zinnatullin It'll be a static method that'll be called at invocation places, so still no additional allocation. If you put the `inline class` instance in a more generic type, it'll automatically be boxed (like `Int` and other primitives are), so it'll call the virtual version of `toString()` and alike methods.

Reply:
How would `System.out.println()` or any other runtime-linked use-site know about any of these methods?

Reply:
@artem-zinnatullin `userId` will be boxed into an instance of `UserId` wrapper class, then this object instance is being passed to `println`

Reply:
aha, I guess it falls into `nullable/platform`-type case.

Still don't like this whole wrapping story, gotta think about it more.